core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php

```diff
@@ -0,0 +1,232 @@
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Core\Exception\BadCredentialsException;
+use Symfony\Component\Security\Core\Security;
+use Symfony\Component\Security\Core\User\PasswordUpgraderInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+use Symfony\Component\Security\Http\Authentication\AuthenticationFailureHandlerInterface;
+use Symfony\Component\Security\Http\Authentication\AuthenticationSuccessHandlerInterface;
+use Symfony\Component\Security\Http\Authenticator\AbstractAuthenticator;
+use Symfony\Component\Security\Http\Authenticator\InteractiveAuthenticatorInterface;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\CsrfTokenBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\PasswordUpgradeBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\RememberMeBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\UserBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Credentials\PasswordCredentials;
+use Symfony\Component\Security\Http\Authenticator\Passport\Passport;
+use Symfony\Component\Security\Http\EntryPoint\AuthenticationEntryPointInterface;
+use Symfony\Component\Security\Http\ParameterBagUtils;
+
+class ContaoLoginAuthenticator extends AbstractAuthenticator implements AuthenticationEntryPointInterface, InteractiveAuthenticatorInterface
+{
+    private array $options;
+
+    public function __construct(
+        private UserProviderInterface $userProvider,
+        private AuthenticationSuccessHandlerInterface $successHandler,
+        private AuthenticationFailureHandlerInterface $failureHandler,
+        private ScopeMatcher $scopeMatcher,
+        private RouterInterface $router,
+        private UriSigner $uriSigner,
+        private ContaoFramework $framework,
+        private TokenStorageInterface $tokenStorage,
+        private PageRegistry $pageRegistry,
+        private HttpKernelInterface $httpKernel,
+        private TwoFactorAuthenticator $twoFactorAuthenticator,
+        array $options,
+    ) {
+        $this->options = array_merge([
+            'username_parameter' => 'username',
+            'password_parameter' => 'password',
+            'check_path' => '/login_check',
+            'post_only' => true,
+            'enable_csrf' => false,
+            'csrf_parameter' => '_csrf_token',
+            'csrf_token_id' => 'authenticate',
+        ], $options);
+    }
+
+    public function start(Request $request, AuthenticationException $authException = null): RedirectResponse|Response
+    {
+        if ($this->scopeMatcher->isBackendRequest($request)) {
+            return $this->redirectToBackend($request);
+        }
+
+        $this->framework->initialize();
+
+        $page = $request->attributes->get('pageModel');
+
+        if (!$page instanceof PageModel) {
+            throw new \RuntimeException('No PageModel found on request.');
```

Comment:
Not sure we can throw an exception here. It is valid to have a controller without a page model inside the front end firewall. Not sure what to do here then though 😬. Maybe not return `true` on `isInteractive`?

Reply:
Adjusted in d2d9bdc98b.

If there is no 401 page configured, it will show an UnauthorizedHttpException:

<img width="1063" alt="Bildschirmfoto 2022-07-06 um 16 53 31" src="https://user-images.githubusercontent.com/754921/177580943-9241c7e7-fa82-4987-9a51-e4ab84ba1dfa.png">